### PR TITLE
lxd: Add generated configuration options to API

### PR DESCRIPTION
**Pull request by gabrielmougard**
This PR replace the following one:
* https://github.com/canonical/lxd/pull/12068

### CODE REVIEW

**Comment by tomponline** [comment]
I think you need to rebase this as you have earlier commits from merged branches here.

File: client/interfaces.go
Comment (tomponline) [inline client/interfaces.go:431]: Why does this need to be added to the client?

File: doc/api-extensions.md
Comment (tomponline) [inline doc/api-extensions.md:2277]: The commit message for this chdnge is out dated

Comment (tomponline) [reply doc/api-extensions.md:2277]: We should change the api extension to make it more general about what the URL endpoint includes as additional content will be added over time.

File: client/interfaces.go
Comment (gabrielmougard) [reply client/interfaces.go:431]: I thought it'd be useful for the LXD UI folk to have a client function to easily get the documentation config

Comment (tomponline) [reply client/interfaces.go:431]: The ui is written in JavaScript so they don't use the go client

Comment (gabrielmougard) [reply client/interfaces.go:431]: Oh ok then I don't think this new function is truly necessary. I can remove it.

Comment (tomponline) [reply client/interfaces.go:431]: Thanks yes please

**Comment by tomponline** [comment]
Happy to merge this once the spelling tests are fixed.

File: lxd/documentation.go
Comment (tomponline) [inline lxd/documentation.go:11]: Given that I dont (yet) see a need for a separate listener for the LXD docs endpoint itself (https://discourse.ubuntu.com/t/expose-the-full-lxd-documentation-website-as-part-of-a-lxd-deployment/37516), and even if we do, we should probably still expose this documentation yaml file over that listener anyway.

So my thinking here is that this `documentation-config-options` endpoint should probably be hierarchically under `/documentation`, but in a way it wont conflict with the full documentation URLs.

Comment (tomponline) [reply lxd/documentation.go:11]: cc @ru-fu

Comment (gabrielmougard) [reply lxd/documentation.go:11]: Yes I agree, it would makes sense to do it this way

Comment (gabrielmougard) [reply lxd/documentation.go:11]: As of now, in [this PR](https://github.com/canonical/lxd/pull/12098), because we have a custom listener for it, we must first a listener address/port (using `core.documentation_address`) and then the doc is accessible at this URL : https:<core.documentation_address>/ . The `/` at the end is not mandatory as there is a redirection rule

Comment (tomponline) [reply lxd/documentation.go:11]: > I'm not sure I agree that this YAML file provides config options documentation.

What would you say it provides? My aim is to avoid adding this long non-hierarchical API endpoint `documentation-config-options`, so open to suggestions.

Comment (ru-fu) [reply lxd/documentation.go:11]: As I said, it provides the available config options. 
I think that for the UI, the list of all options and their grouping will be more important than the documentation for them.

Comment (tomponline) [reply lxd/documentation.go:11]: What is your proposal for api endpoint URL?

Comment (ru-fu) [reply lxd/documentation.go:11]: Sounds good to me!

Comment (tomponline) [reply lxd/documentation.go:11]: Thanks. @gabrielmougard please can you go ahead and make that change as a priority so we can get this merged for LXD 5.17. Thanks

File: doc/api-extensions.md
Comment (tomponline) [inline doc/api-extensions.md:2280]: This API extension should change to `metadata_configuration`

File: Makefile
Comment (tomponline) [inline Makefile:116]: This should become update-metadata

File: Makefile
Comment (tomponline) [inline Makefile:119]: Lets use lxd/metadata/configuration.yaml for this file

File: lxd/api_1.0.go
Comment (tomponline) [inline lxd/api_1.0.go:52]: metdataConfigurationCmd

File: lxd/documentation.go
Comment (gabrielmougard) [reply lxd/documentation.go:11]: sure

**Comment by tomponline** [comment]
@gabrielmougard this can be rebased now.

File: lxd/metadata/configuration.yaml
Comment (tomponline) [inline lxd/metadata/configuration.yaml:1]: Comment being added is incorrect

**Review by tomponline — APPROVED** [review]

File: lxd/metadata/configuration.yaml
Comment (ru-fu) [reply lxd/metadata/configuration.yaml:711]: The backticks turn them into code in Markdown. And the quotes are needed to make it valid YAML afair.

Comment (ru-fu) [reply lxd/metadata/configuration.yaml:711]: Different question: Why is this file checked in? Shouldn't it be generated dynamically?

Comment (tomponline) [reply lxd/metadata/configuration.yaml:711]: > Different question: Why is this file checked in? Shouldn't it be generated dynamically?

I understand @gabrielmougard wanted it to be generated by the developer and committed, like we do with the API docs, and i18n translations.

Comment (tomponline) [reply lxd/metadata/configuration.yaml:711]: > The generated `.txt` document (used for the site documentation), is derived from the YAML document (the marshaled YAML is the 'source of truth') in the code.

What tool does this conversion? It seems to me that this tool should be the one converting it to markdown, rather than being double encoded in the yaml file that is also served out through the API for non-documentation consumers (e.g. UI).

Comment (ru-fu) [reply lxd/metadata/configuration.yaml:711]: There is a semantic difference, so I think it should be in the code comment.
Think about a default value - "no" could mean that there is no default value, while "`no`" means the default value is `no`.

Comment (ru-fu) [reply lxd/metadata/configuration.yaml:711]: The "no" was just one example. It could also be "auto" - which can mean either the value `auto` or that the value is automatically determined.

Forget about what I said about Markdown. It's really a differentiation between a specific value and a description.

Comment (ru-fu) [reply lxd/metadata/configuration.yaml:711]: We also have things like ``default: `10` (maximum)``, for example, which has both a value and a description.

Comment (tomponline) [reply lxd/metadata/configuration.yaml:711]: @ru-fu ah OK I see the challenge.

Because this is a metadata API rather than purely for documentation generation, it sounds to me like we need to come up with a slightly different schema that includes additional fields that can enable us to describe programmatically what the default values are and what the valid values/ranges are for a field (at a basic level).

So to check I understand the things we're trying to represent here, they are:

- What type of default value it has (potentially a list of defined values that indicate whether there is a default at all or if the value is automatically determined).
- The static default value (if there is one) - which would be its literal value. A question I have here is whether we should use YAML's built in data types to represent the value (i.e boolean for true/false, int for numbers, strings for text etc) or whether we should always present as a string (because that is what the API will accept for configuration items).
- One or more optional fields that allow describing the basic validation of the field (i.e list of valid literal values or ranges of numbers).

This sounds a bit like the schema we designed for ACL rules, but more general:

https://documentation.ubuntu.com/lxd/en/latest/howto/network_acls/#rule-properties

Then the tool that converts the yaml file to the documentation text representation would have the knowledge to take the metadata and convert it to the correct representation for the documentation. And the UI would be able to consume the YAML via the API and generate the form fields (with basic validation) from it.

Comment (ru-fu) [reply lxd/metadata/configuration.yaml:711]: I guess that's the discussion with the UI team that needs to happen.

From a documentation perspective, this might be overshooting it a bit - we just need a series of fields/keys and their content, similar to the tables we had before (and we don't only have the "default" field, but several others - the ones that are currently supported are here: https://github.com/canonical/lxd-sphinx-extensions/blob/9a52e049fd2de70e0451cc725ba7ed2669afbd7c/config-options/__init__.py#L27 ). But as long as the tool generated suitable doc blocks, the input can be more complicated if needed.

Comment (tomponline) [reply lxd/metadata/configuration.yaml:711]: @gabrielmougard with the above discussion in mind please can you have a think about possible approaches here and update the spec post so we can have something to discuss with the UI team. Thanks

Comment (MusicDin) [reply lxd/metadata/configuration.yaml:711]: Hi everyone,

I'm wondering whether there is a way for us to expand the schema also for storage pools?

Also, would it make sense to have a field `managed` (or something like this) in the schema indicating that if the field is not set, it will be managed by the LXD itself. For some settings this cannot be represented with the default value. For example, IP address does not have a default value, but the value will be set by the LXD (if possible of course).

Example use case:

Having this for storage (potentially for all storage drivers) would allow external applications to know ahead of time which settings will be set by LXD. To be more specific, in `terraform-lxd-provider` there is currently no way to detect which settings are managed by the LXD, causing inconsistencies between Terraform runs (see also terraform-lxd/terraform-provider-lxd/issues/328). Or in other words, the user is required to always provide ceratin settings, such as pool's source and size, otherwise Terraform will try to delete those fields when the configuration is reapplied.

Comment (MusicDin) [reply lxd/metadata/configuration.yaml:711]: Seems like this is already in the schema - I've somehow missed the link provided by @ru-fu: https://github.com/canonical/lxd-sphinx-extensions/blob/9a52e049fd2de70e0451cc725ba7ed2669afbd7c/config-options/__init__.py#L27

Comment (ru-fu) [reply lxd/metadata/configuration.yaml:711]: Yes, from a documentation perspective, we want to use this tool to generate all configuration option tables, including the ones for the storage pools.
Here's the list of which are still missing: https://warthogs.atlassian.net/browse/LXD-430

Comment (MusicDin) [reply lxd/metadata/configuration.yaml:711]: Thanks for the link, can I help with some subtasks?

Comment (tomponline) [reply lxd/metadata/configuration.yaml:711]: We should probably wait for @gabrielmougard's updated spec proposal on schema before we start updating the code itself as it might mean additional rework later. We're also due to have a meeting with the UX team to discuss their needs.

File: lxd/metadata/configuration.yaml
Comment (ru-fu) [inline lxd/metadata/configuration.yaml:1]: Bringing this up as a separate comment again since it's not related to the other discussion.

> > Different question: Why is this file checked in? Shouldn't it be generated dynamically?
> 
> I understand @gabrielmougard wanted it to be generated by the developer and committed, like we do with the API docs, and i18n translations.

If we check in this file, shouldn't we also check in the generated documentation file and get rid of the extra code to build the go tool on Read the docs?

Comment (tomponline) [reply lxd/metadata/configuration.yaml:1]: Hi @ru-fu as I'm not totally familiar with the process you and @gabrielmougard have setup for this work I'll try and give some guidelines as to what we need and what we don't want to end up with, and hopefully we can arrive at a solution that works for all. But @gabrielmougard please do correct me if I'm mistaken with any of this.

Because the yaml file (only) is part of the LXD API, it needs to be embedded into the LXD Go binary itself, so that it can be available without external dependencies (i.e without distributing that file as well).

For this @gabrielmougard is using a technique which I understand requires the yaml file to exist before LXD can be built, so it can be embedded. So we've followed a similar approach as we do with the DB generation stuff, by using a build-time step to generate the yaml file and then a github check to make sure its kept up-to-date with changes in the source code.

So I'm happy that what we are doing with the yaml file is correct and consistent with how we've done other similar items.

Now, onto your question: 

> If we check in this file, shouldn't we also check in the generated documentation file and get rid of the extra code to build the go tool on Read the docs?

This is where things start to get a bit hazy for me.

I think you're referring to this PR https://github.com/canonical/lxd/pull/12043

Are you saying that the docs we store in https://github.com/canonical/lxd/tree/main/doc are now no longer "complete" in that they are missing the configuration options, and that these parts of the documentation only get built in readthedocs?

If so, then I think I would agree with you that we should be building and committing the dynamic part of the docs into the repo so that the complete docs are available in the githubrepo and not dependent on a build step in readthedocs?

But its possible I've not considered something that you and/or @gabrielmougard wanted to achieve with this originally.

File: lxd/metadata/configuration.yaml
Comment (tomponline) [reply lxd/metadata/configuration.yaml:711]: I would only say "growing" WRT to @MusicDin , but the original requirement has always been for the UI to consume the YAML file via the API programmatically, and so in that sense the original requirements remain unchanged.

I don't think we need anything "private", but what we need is to remove any markdown encoding in the YAML response.

In doing this we will encounter the problem that currently the markdown in the `default` field is encoding multiple pieces of information in human readable form, and its that which is problematic here.

Instead we should introduce more fields to the YAML file that allow for encoding the information needed to *both* generate the resulting documentation markdown (in the current format it is) *and* to allow the UX and Terraform providers to be able to parse the information about each field programmatically.

Comment (MusicDin) [reply lxd/metadata/configuration.yaml:711]: I like your approach, however, to me it would seem cleaner to have a `default` field and `defaultDoc`, `defaultDisplay`, or maybe simply `display`. This is just a thought, but this way, the difference between the fields would be clear from their names.

Comment (tomponline) [reply lxd/metadata/configuration.yaml:711]: I'd like to try and avoid duplicating the default info if we can. Really the markdown values being put in the default field at the moment are expressing things beyond the default value, such as validation rules. So rather than duplicating, we should try and find a way to express the validation info in a separate field.

I believe we could also use the absence of a default field to indicate if there is "no" default value.

File: lxd/metadata/configuration.yaml
Comment (gabrielmougard) [reply lxd/metadata/configuration.yaml:1]: > Because the yaml file (only) is part of the LXD API, it needs to be embedded into the LXD Go binary itself, so that it can be available without external dependencies (i.e without distributing that file as well).

> For this @gabrielmougard is using a technique which I understand requires the yaml file to exist before LXD can be built, so it can be embedded. So we've followed a similar approach as we do with the DB generation stuff, by using a build-time step to generate the yaml file and then a github check to make sure its kept up-to-date with changes in the source code.

That is correct. As for the readthedoc hook, I think it'd be more consistent to check the .txt file in (as we do for the yaml. I don't like the idea to decouple these two files in two different build steps as they serve a similar purpose on some aspects). But this is just a very personal design point of view.

Comment (tomponline) [reply lxd/metadata/configuration.yaml:1]: OK, and how does the .txt file end up in the resulting docs?

Comment (tomponline) [reply lxd/metadata/configuration.yaml:1]: > I don't like the idea to decouple these two files in two different build steps as they serve a similar purpose on some aspects). 

Out of interest, why did we do this then?

Comment (ru-fu) [reply lxd/metadata/configuration.yaml:1]: > Are you saying that the docs we store in https://github.com/canonical/lxd/tree/main/doc are now no longer "complete" in that they are missing the configuration options, 

Correct - to be able to build the docs, we need the `config_options.txt` that is generated by the `lxd-metadata` tool.

> and that these parts of the documentation only get built in readthedocs?

No, fortunately not. ;)
We originally added building the `config_options.txt` file to the Makefile (see https://github.com/canonical/lxd/pull/11652). I was a bit sceptical to this since I find it overkill to require Go for building the docs, but it was the only option at that time since we didn't want to require this file to be checked in.
When we moved to RTD, we had to move this functionality to the actual doc build (`doc/conf.py`), because RTD doesn't support Makefiles. It still runs for local builds as well though.
Still, I don't think we should misuse RTD to build Go tools, so if we now require the YAML file created by the same tool to be checked in, we should do the same for the TXT file for documentation.

File: lxd/metadata/configuration.yaml
Comment (ru-fu) [reply lxd/metadata/configuration.yaml:711]: I agree that we should not duplicate any information.
Remember that you'll need to write these comments for every new config option you add, and you don't want to write the same information in two different formats.

File: lxd/metadata/configuration.yaml
Comment (tomponline) [reply lxd/metadata/configuration.yaml:1]: Thanks. 

Can you show me how we go from raw source code to complete docs locally at the moment?

Comment (ru-fu) [reply lxd/metadata/configuration.yaml:1]: `config_options.txt` is generated through the [generate_go_docs](https://github.com/canonical/lxd/blob/2aca8195ba21eb12b1cf937f6b28d867676600b2/doc/conf.py#L202) Sphinx extension.

Sections of the file are then included in other documentation files, for example [here](https://github.com/canonical/lxd/blob/2aca8195ba21eb12b1cf937f6b28d867676600b2/doc/server.md?plain=1#L29)

Comment (ru-fu) [reply lxd/metadata/configuration.yaml:1]: Oh, and the last piece of the puzzle is the [config-options Sphinx extension](https://github.com/canonical/lxd-sphinx-extensions/blob/main/config-options/__init__.py) which tells Sphinx how to deal with the custom Sphinx directives in `config_options.txt`. Just mentioning that because it's also something we have made and can change at any time if we need updates there to get stuff to show up as we want.

File: Makefile
Comment (ru-fu) [inline Makefile:119]: This is the same as generate-config below - we only need it once. ;)

File: lxd/lxd-metadata/lxddoc.go
Comment (ru-fu) [inline lxd/lxd-metadata/lxddoc.go:1]: Can you update the readme file with the new name as well? That one still says lxd-doc.

**Comment by tomponline** [comment]
@gabrielmougard please can you check all tests pass and that you have replaced all references to YAML with JSON (there are some you missed). Thanks

**Comment by ru-fu** [comment]
Here's the PR that adds the "defaultdesc" field to the config options Sphinx extension: https://github.com/canonical/lxd-sphinx-extensions/pull/20

To use this version of the extension when building the docs here in this PR, replace `lxd-sphinx-extensions` in `doc/.sphinx/requirements.txt` with:
```
git+https://github.com/canonical/lxd-sphinx-extensions@refs/pull/20/merge#egg=lxd-sphinx-extensions
```

**Comment by ru-fu** [comment]
Are you translating "defaultdesc" to "default" for the doc output now? I don't think that's a good idea, it'll lead to confusion.

File: lxd/lxd-metadata/README.md
Comment (tomponline) [inline lxd/lxd-metadata/README.md:94]: This example is missing defaultdest and has default instead.

File: lxd/lxd-metadata/README.md
Comment (tomponline) [inline lxd/lxd-metadata/README.md:104]: defaultdesc

File: lxd/documentation.go
Comment (tomponline) [inline lxd/documentation.go:60]: We should parse the json on disk into a struct and then return it to the client using `response.SyncResponse(true, data)` that way it will be returned using a context type of "application/json" and will be wrapped in the standard envelope for our JSON API:

```
 {
		"type": "sync",
		"status": "Success",
		"status_code": 200,
		"operation": "",
		"error_code": 0,
		"error": "",
		"metadata": {
...
}
}
```

File: doc/api-extensions.md
Comment (tomponline) [inline doc/api-extensions.md:2286]: YAML?

Comment (tomponline) [reply doc/api-extensions.md:2286]: Also, the API extension should describe the resulting API changes to the end user, rather than implementation.
So there's no need to mention the "file containing the metadata" - the end user doesn't need to know that there is a file at all, only that the new URL endpoint returns metadata about configuration in a certain format.

**Review by tomponline — CHANGES_REQUESTED** [review]

**Comment by gabrielmougard** [comment]
> Here's the PR that adds the "defaultdesc" field to the config options Sphinx extension: [canonical/lxd-sphinx-extensions#20](https://github.com/canonical/lxd-sphinx-extensions/pull/20)
> 
> To use this version of the extension when building the docs here in this PR, replace `lxd-sphinx-extensions` in `doc/.sphinx/requirements.txt` with:
> 
> ```
> git+https://github.com/canonical/lxd-sphinx-extensions@refs/pull/20/merge#egg=lxd-sphinx-extensions
> ```

@ru-fu I have an error saying `couldn't find remote ref refs/pull/20/merge` when I try to do this

**Comment by ru-fu** [comment]
Try 
```
git+https://github.com/canonical/lxd-sphinx-extensions@refs/pull/21/merge#egg=lxd-sphinx-extensions
```
now.

**Comment by gabrielmougard** [comment]
> ```
> git+https://github.com/canonical/lxd-sphinx-extensions@refs/pull/21/merge#egg=lxd-sphinx-extensions
> ```

It works locally, let see is the CI is happy...

File: doc/api-extensions.md
Comment (ru-fu) [inline doc/api-extensions.md:2286]: This is way too specific - these are the entities and sections that we have right now, but we will add others.

Comment (gabrielmougard) [reply doc/api-extensions.md:2286]: What do you think about 
```md
Adds `GET /1.0/metadata/configuration` API endpoint to retrieve the generated metadata configuration in a JSON format for all the cluster, instance and server options. The JSON structure adopts the structure ```"configs" > `ENTITY` > `ENTITY_SECTION` > "keys" > [<CONFIG_OPTION_0>, <CONFIG_OPTION_1>, ...]```
```

Comment (ru-fu) [reply doc/api-extensions.md:2286]: "for all the cluster, instance and server options" - this is right now. We want to use it for everything eventually.
If you want to make it clear that we're not covering everything yet, you could link to https://documentation.ubuntu.com/lxd/en/latest/config-options/ which has a list of options that are currently treated like this.
But we will never remember to go back here and update the description when we add more options (plus, should any of these descriptions be updated at all?).

**Review by MusicDin — APPROVED** [review]

File: doc/api-extensions.md
Comment (ru-fu) [inline doc/api-extensions.md:2286]:
```suggestion
Adds the `GET /1.0/metadata/configuration` API endpoint to retrieve the generated metadata configuration in a JSON format. The JSON structure adopts the structure ```"configs" > `ENTITY` > `ENTITY_SECTION` > "keys" > [<CONFIG_OPTION_0>, <CONFIG_OPTION_1>, ...]```. 
Check the list of {ref}`config-options` to see which configuration options are included.
```

File: doc/.sphinx/requirements.txt
Comment (ru-fu) [inline doc/.sphinx/requirements.txt:37]: Need to revert this again before merging.

**Review by ru-fu — CHANGES_REQUESTED** [review]

File: doc/api-extensions.md
Comment (ru-fu) [inline doc/api-extensions.md:2286]: Why would you mark `configuration options` as code? My suggestion should work as is.
I also suggested a "the" in the first sentence.

Comment (ru-fu) [reply doc/api-extensions.md:2286]: Also, there's no information about the structure in that link.

Comment (ru-fu) [reply doc/api-extensions.md]: > Why would you mark `configuration options` as code? My suggestion should work as is. I also suggested a "the" in the first sentence.

Interesting - the link works fine in incremental builds, but not when building from scratch.
This should do it: 

```
{doc}`configuration options </config-options>`
```

**Review by ru-fu — APPROVED** [review]
LGTM!

**Pull request merged**
